### PR TITLE
Dilate additions

### PR DESCRIPTION
**Pull request by dematteojesse**
Very big addition. Incorporating trial input loading data, dilate/softdtw loss functions and evaluation metrics, label modifications, data saving capabilities (including normalization scalers), split into validation set. May be some minor adjustments that cause breaking changes to existing code.

### CODE REVIEW

File: neuraldecoding/dataaugmentation/__init__.py
Comment (lhcubillos) [inline neuraldecoding/dataaugmentation/__init__.py:1]: The _init_ files should keep the relative import, as that is what guides the imports of the library

File: neuraldecoding/dataset/dataset.py [hunk 1]
```diff
@@ -64,8 +63,8 @@ def _load_data_nwb(self):
         Load data from NWB file
 
         """
-        self.io = NWBHDF5IO(self.cfg.dataset_parameters.nwb_file, mode="r")
```
Comment (lhcubillos) [inline neuraldecoding/dataset/dataset.py:67]: This is important, we shouldn't remove it

File: neuraldecoding/dataset/dataset.py [hunk 2]
```diff
@@ -133,18 +134,4 @@ def save_data(self):
             path = self.cfg.save_path
         print(f"Saving NWB file to {path}...")
         with NWBHDF5IO(path, mode="w") as io:
-            io.write(self.dataset)
-
-    def close(self):
```
Comment (lhcubillos) [inline neuraldecoding/dataset/dataset.py:138]: we shouldn't remove any of the stuff through the bottom either

File: neuraldecoding/dataset/dataset.py (hunk 1 above)
Comment (dematteojesse) [reply neuraldecoding/dataset/dataset.py:67]: why?

File: neuraldecoding/dataset/dataset.py (hunk 2 above)
Comment (dematteojesse) [reply neuraldecoding/dataset/dataset.py:138]: yeah follows from the first thing

File: neuraldecoding/dataset/dataset.py (hunk 1 above)
Comment (lhcubillos) [reply neuraldecoding/dataset/dataset.py:67]: We need to keep the file open while we do any stuff with the data, because it doesn't actually copy all the contents to memory. The self.io object helps with doing that, and the rest of the methods are to make sure we close the file when the dataset object gets deleted

File: neuraldecoding/preprocessing/blocks.py [hunk 3]
```diff
@@ -86,67 +88,42 @@ def transform(self, data, interpipe):
 			interpipe (dict): A inter-pipeline bus for one-way sharing data between blocks within the preprocess_pipeline call.
 		"""
 		pass
-
-
-# Wrappers that Modifies Data Format
-class Dict2DataDictBlock(DataFormattingBlock):
```
Comment (lhcubillos) [inline neuraldecoding/preprocessing/blocks.py:92]: Why did you delete this? I believe people are using it

File: neuraldecoding/preprocessing/blocks.py [hunk 4]
```diff
@@ -750,6 +741,82 @@ def transform(self, data, interpipe):
...
+		if isinstance(locs, str):
+			locs = [locs]
+		if isinstance(keys, str):
+			keys = [keys]
+		if len(locs) != len(keys):
+			raise ValueError("locs and keys must be the same length.")
+		self.locs = locs
+		self.keys = keys
+	
+	def transform(self, data, interpipe):
```
Comment (lhcubillos) [inline neuraldecoding/preprocessing/blocks.py:812]: It might be worth just saving the full data and interpipe dictionaries into a single pkl file that makes it easy to restore the state of the pipeline at this point

File: neuraldecoding/feature_extraction/feature_extraction.py
[the remark this answers is not in the document]
Comment (dematteojesse) [reply neuraldecoding/feature_extraction/feature_extraction.py:1]: Bc of working around the return_array parameter. I can't get the start of each bin if the output is an array. I think deleting the return_array if block might be good bc it seems like in the latest version of main, you handle converting to an array outside of the function anyway

**Review by dematteojesse — COMMENTED** [review]
All changes addressed, check what you think about the removing "return_array" param from the feature extractor. also lmk if you want me to replace the datadict2dict block

**Comment by htemmar** [comment]
pull main again to make sure
